From c707b5f3c8e3c7b98b772d3b75c106850bcc651e Mon Sep 17 00:00:00 2001
From: Tanuj Tekriwal <tanuj.tekriwal@intel.com>
Date: Tue, 26 Nov 2019 14:30:27 +0530
Subject: [PATCH] driver:usb:role: roles: Fix USB 3.0 OTG issue on Intel
 platform

This patch adds static DRD mode for host/device mode switch. This fixes the issue where device mode was not working after DUT switches to host mode with 3.0 OTG connector.

Signed-off-by: Tanuj Tekriwal <tanuj.tekriwal@intel.com>
---
 drivers/usb/roles/intel-xhci-usb-role-switch.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/drivers/usb/roles/intel-xhci-usb-role-switch.c b/drivers/usb/roles/intel-xhci-usb-role-switch.c
index 277de96181f9..b90053be745e 100644
--- a/drivers/usb/roles/intel-xhci-usb-role-switch.c
+++ b/drivers/usb/roles/intel-xhci-usb-role-switch.c
@@ -26,6 +26,9 @@
 #define SW_VBUS_VALID			BIT(24)
 #define SW_IDPIN_EN			BIT(21)
 #define SW_IDPIN			BIT(20)
+#define SW_SWITCH_EN_CFG0		BIT(16)
+#define SW_DRD_STATIC_HOST_CFG0		1
+#define SW_DRD_STATIC_DEV_CFG0		2
 
 #define DUAL_ROLE_CFG1			0x6c
 #define HOST_MODE			BIT(29)
@@ -65,17 +68,22 @@ static int intel_xhci_usb_set_role(struct device *dev, enum usb_role role)
 	case USB_ROLE_NONE:
 		val |= SW_IDPIN;
 		val &= ~SW_VBUS_VALID;
+		val &= ~(SW_DRD_STATIC_DEV_CFG0 | SW_DRD_STATIC_HOST_CFG0);
 		break;
 	case USB_ROLE_HOST:
 		val &= ~SW_IDPIN;
 		val &= ~SW_VBUS_VALID;
+		val &= ~SW_DRD_STATIC_DEV_CFG0;
+		val |= SW_DRD_STATIC_HOST_CFG0;
 		break;
 	case USB_ROLE_DEVICE:
 		val |= SW_IDPIN;
 		val |= SW_VBUS_VALID;
+		val &= ~SW_DRD_STATIC_HOST_CFG0;
+		val |= SW_DRD_STATIC_DEV_CFG0;
 		break;
 	}
-	val |= SW_IDPIN_EN;
+	val |= SW_IDPIN_EN | SW_SWITCH_EN_CFG0;
 
 	writel(val, data->base + DUAL_ROLE_CFG0);
 
-- 
2.17.1

